ui/src/routes/devices.$id.settings.macros.tsx

```diff
@@ -0,0 +1,1439 @@
+import { useState, useEffect, useRef, useCallback } from "react";
+import { LuPlus, LuTrash, LuX, LuPenLine, LuLoader, LuGripVertical, LuInfo } from "react-icons/lu";
+import { Combobox, ComboboxInput, ComboboxOption, ComboboxOptions } from "@headlessui/react";
+
+import { KeySequence, useMacrosStore } from "../hooks/stores";
+import { SettingsPageHeader } from "../components/SettingsPageheader";
+import { Button } from "../components/Button";
+import { keys, modifiers } from "../keyboardMappings";
+import { useJsonRpc } from "../hooks/useJsonRpc";
+import notifications from "../notifications";
+import { SettingsItem } from "../routes/devices.$id.settings";
+
+const DEFAULT_DELAY = 50;
+
+interface MacroStep {
+  keys: string[];
+  modifiers: string[];
+  delay: number;
+}
+
+interface KeyOption {
+  value: string;
+  label: string;
+}
+
+interface KeyOptionData {
+  value: string | null;
+  keys?: string[];
+  label?: string;
+}
+
+const generateId = () => {
+  return `macro-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`;
+};
+
+const keyOptions = Object.keys(keys).map(key => ({
+  value: key,
+  label: key,
+}));
+
+const modifierOptions = Object.keys(modifiers).map(modifier => ({
+  value: modifier,
+  label: modifier.replace(/^(Control|Alt|Shift|Meta)(Left|Right)$/, "$1 $2"),
+}));
+
+const groupedModifiers = {
+  Control: modifierOptions.filter(mod => mod.value.startsWith('Control')),
+  Shift: modifierOptions.filter(mod => mod.value.startsWith('Shift')),
+  Alt: modifierOptions.filter(mod => mod.value.startsWith('Alt')),
+  Meta: modifierOptions.filter(mod => mod.value.startsWith('Meta')),
+};
+
+interface KeyComboboxProps {
+  stepIndex: number;
+  step: MacroStep;
+  onSelect: (option: KeyOptionData) => void;
+  query: string;
+  onQueryChange: (query: string) => void;
+  getFilteredOptions: () => KeyOption[];
+  disabled?: boolean;
+}
+
+function KeyCombobox({
```

Comment:
Look at the existing Input Field component or Select Menu, and adopt the same interface and styling. This should feel like that, just have a "dropdown". We should be able to use this elsewhere too.

Reply:
I redid this component into Combobox. tried to style as similarly as i could